src/shortest_path_graph.py

```diff
@@ -0,0 +1,206 @@
+                if edge[0] == node_delete:
+                    self.graph[node].remove(edge)
+
+    def del_edge(self, node1, node2):
+        """Delete the edge connecting node1 to node 2 if it exists."""
+        if node2 not in [edge[0] for edge in self.graph[node1]]:
+            raise IndexError('This edge does not exist.')
+
+        removed_edge = [edge for edge in self.graph[node1] if edge[0] == node2]
+        self.graph[node1].remove(removed_edge[0])
+
+    def has_node(self, node):
+        """Return true if the input node is in the graph, else False."""
+        return node in self.graph
+
+    def neighbours(self, node):
+        """Return the list of nodes connected to the input node."""
+        return [edge[0] for edge in self.graph[node]]
+
+    def adjacent(self, node1, node2):
+        """Return True if there is an edge connecting n1 and n2."""
+        if node1 not in self.graph or node2 not in self.graph:
+            raise KeyError('One or both of these nodes is not in the graph.')
+        return node2 in [edge[0] for edge in self.graph[node1]]
+
+    def depth_traversal(self, root, visited=None):
+        """Perform depth traversal of graph."""
+        if visited is None:
+            visited = []
+        visited.append(root)
+        for edge in self.graph[root]:
+            if edge[0] not in visited:
+                self.depth_traversal(edge[0], visited)
+        return visited
+
+    def breadth_traversal(self, root):
+        """Perform breath traversal of graph."""
+        visited = [root]
+        node_edges = self.graph[root]
+        while node_edges:
+            edge = node_edges.pop(0)
+            if edge[0] not in visited:
+                visited.append(edge[0])
+                unique_edges = [edge[0] for edge in self.graph[edge]
+                                if edge[0] not in visited]
+                node_edges.extend(unique_edges)
+        return visited
+
+    def dijkstra(self, source, target):
+        """Find the shorted path from source node to all other nodes."""
+        visited = {source: 0}
+        path = {}
+        nodes_visit = self.nodes()
+        while nodes_visit:
+            min_node = None
+            for node in nodes_visit:
+                if node in visited:
+                    if min_node is None:
+                        min_node = node
+                    elif visited[node] < visited[min_node]:
+                        min_node = node
+            if min_node is None:
+                break
+
+            nodes_visit.remove(min_node)
+            current_weight = visited[min_node]
+
+            for edge in self.graph[min_node]:
+                weight = current_weight + edge[1]
+                if edge[0] not in visited or weight < visited[edge[0]]:
+                    visited[edge[0]] = weight
+                    path[edge[0]] = min_node
+
+            if min_node == target:
+                return visited[target], self._path(source, target, path)
+
```

Comment:
This is really good.  It should also raise a ValueError though when there is not path from start to target